core/drivers/clk/sam/at91_cpu_opp.c

```diff
@@ -18,6 +18,44 @@ struct clk_rates {
 
 static struct clk_rates *opp_rates;
 
+static TEE_Result get_rates_array(struct clk *clk __unused, size_t start_index,
+				  unsigned long *rates, size_t *nb_elts)
+{
+	if (!opp_rates)
+		panic("Invalid CPU OPP Rates Array");
+
+	if (!rates) {
+		*nb_elts = opp_rates->rate_num;
+
+		return TEE_SUCCESS;
+	}
+
+	if (start_index > opp_rates->rate_num ||
+	    start_index + *nb_elts > opp_rates->rate_num) {
+		EMSG("Bad parameter(s): start_index %d, nb_elts %d",
+		     start_index, *nb_elts);
+
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	memcpy(rates, &opp_rates->rates[start_index],
+	       *nb_elts * sizeof(*rates));
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result cpu_opp_clk_set_rate(struct clk *clk, unsigned long rate,
+				       unsigned long parent_rate)
+{
+	assert(clk->parent);
+
```

Comment:
I would recommend here to check target `parent_rate` is lists in `opp_rates->rates[]`.

Reply:
Updated according to the comments except this one, do you mean to check `parent_rate` with the following patch? Thanks.
```
--- a/core/drivers/clk/sam/at91_cpu_opp.c
+++ b/core/drivers/clk/sam/at91_cpu_opp.c
@@ -14,6 +14,7 @@

 /* the clock rates for CPU OPP */
 struct clk_rates {
+       unsigned long parent_rate;
        size_t rate_num; /* the number of valid clock rates in @rates */
        unsigned long rates[OPP_RATES_MAX];
 };
@@ -48,6 +49,7 @@ static TEE_Result cpu_opp_clk_set_rate(struct clk *clk, unsigned long rate,
                                       unsigned long parent_rate)
 {
        assert(clk->parent);
+       assert(opp_rates->parent_rate == parent_rate);

        return clk->parent->ops->set_rate(clk->parent, rate, parent_rate);
 }
@@ -132,6 +134,7 @@ TEE_Result at91_clk_register_cpu_opp(const void *fdt, int node, struct clk *clk)
        if (res)
                return res;

+       opp_rates->parent_rate = clk_get_rate(clk);
        cpu_opp_clk = clk_alloc("cpu-opp", &cpu_opp_clk_ops, &clk, 1);
        if (!cpu_opp_clk)
                panic("CPU OPP clock alloc failed");
```

Reply:
No, I was rather thinking of something like:
```c
for (n = 0; n < opp_rates->rate_num; n++)
    if (rate == opp_rates->rates[n])
        break;
if (n == opp_rates->rate_num)
    return TEE_ERROR_BAD_PARAMETERS;

Reply:
The `rate` check added, thank you.